File: php/Factory/SimpleFactory.php
<?php

class Factory {

    static function createIM($type) :IM {
        switch ($type) {
            case "wechat":
                return new WeChat();
            case "feishu":
                return new Feishu();
        }
        throw new Exception("Invalid type");
    }
}

interface IM {
    function send($msg);
}

class WeChat implements IM {
    function send($msg)
    {
        echo "Send msg by Wechat: {$msg} \n";
    }
}

class Feishu implements  IM {
    function send($msg)
    {
        echo "Send msg by Feishu: {$msg} \n";
    }
}

$im = Factory::createIM("wechat");
$im->send("hello ?");

File: go/src/Factory/SimpleFactory.go
package Factory

import "fmt"

func Execute() {
	im := createIm("feishu")
	im.Send("hello ?")
}

type IM interface {
	Send(msg string)
}

type Wechat struct{}

type Feishu struct{}

func (f *Feishu) Send(msg string) {
	fmt.Println("Send msg by feishu:" + msg)
}

func (w *Wechat) Send(msg string) {
	fmt.Println("Send msg by wechat:" + msg)
}

func createIm(t string) IM {
	switch t {
	case "wechat":
		return &Wechat{}
	case "feishu":
		return &Feishu{}
	}
	panic("invalid im type")
}
